src/expr/src/interpret.rs

```diff
@@ -585,6 +585,111 @@ impl From<Monotonic> for Pushdownable {
+        /// on values.
+        fn eagerly<'b>(
+            left: ResultSpec<'b>,
+            right: ResultSpec<'b>,
+            value_fn: impl FnOnce(Values<'b>, Values<'b>) -> ResultSpec<'b>,
+        ) -> ResultSpec<'b> {
+            let result = match (left.values, right.values) {
+                (Values::Empty, _) | (_, Values::Empty) => ResultSpec::nothing(),
+                (l, r) => value_fn(l, r),
+            };
+            ResultSpec {
+                fallible: left.fallible || right.fallible || result.fallible,
+                nullable: left.nullable || right.nullable || result.nullable,
+                values: result.values,
+            }
+        }
+
+        fn jsonb_get_string<'b>(
+            left: ResultSpec<'b>,
+            right: ResultSpec<'b>,
+            stringify: bool,
+        ) -> ResultSpec<'b> {
+            eagerly(left, right, |left, right| {
+                let nested_spec = match (left, right) {
+                    (Values::Nested(mut map_spec), Values::Within(key, key2)) if key == key2 => {
+                        map_spec.remove(&key)
+                    }
+                    _ => None,
+                };
+
+                if let Some(field_spec) = nested_spec {
+                    if stringify {
+                        // We only preserve value-range information when stringification
+                        // is a noop. (Common in real queries.)
+                        let values = match field_spec.values {
+                            Values::Empty => Values::Empty,
+                            Values::Within(min @ Datum::String(_), max @ Datum::String(_)) => {
+                                Values::Within(min, max)
+                            }
+                            Values::Within(_, _) | Values::Nested(_) | Values::All => Values::All,
+                        };
+                        ResultSpec {
+                            values,
+                            ..field_spec
+                        }
+                    } else {
+                        field_spec
+                    }
+                } else {
+                    // TODO: it should be possible to narrow this further...
+                    ResultSpec::anything()
+                }
+            })
+        }
+
+        fn eq<'b>(left: ResultSpec<'b>, right: ResultSpec<'b>) -> ResultSpec<'b> {
+            eagerly(left, right, |left, right| {
+                // `eq` might return true if there's any overlap between the range of its two arguments...
+                let maybe_true = match left.clone().intersect(right.clone()) {
+                    Values::Empty => ResultSpec::nothing(),
+                    _ => ResultSpec::value(Datum::True),
+                };
+
+                // ...and may return false if the union contains at least two distinct values.
+                let maybe_false = match left.union(right) {
```

Comment:
Maybe add a comment here that if either side is `Values::Empty`, that is handled by `eagerly`.